File: fast-fibonacci-algorithms/fastFibonacci.hs
{- 
 - Fast doubling Fibonacci algorithm
 - 
 - Copyright (c) 2014 Project Nayuki
 - All rights reserved. Contact Nayuki for licensing.
 - http://www.nayuki.io/page/fast-fibonacci-algorithms
 -}


-- fibonacci n = F(n)
fibonacci :: Integer -> Integer
fibonacci n | n >= 0 = a where (a, _) = fib n

-- Internal function: fib n = (F(n), F(n+1))
fib :: Integer -> (Integer, Integer)
fib 0 = (0, 1)
fib n =
	let (a, b) = fib (div n 2)
		c = a * (2 * b - a)
		d = b * b + a * a
	in if mod n 2 == 0
		then (c, d)
		else (d, c + d)

File: fast-fibonacci-algorithms/fastfibonacci.py
# 
# Fast doubling Fibonacci algorithm
# 
# Copyright (c) 2015 Project Nayuki
# All rights reserved. Contact Nayuki for licensing.
# http://www.nayuki.io/page/fast-fibonacci-algorithms
# 


# (Public) Returns F(n).
def fibonacci(n):
    if n < 0:
        raise ValueError("Negative arguments not implemented")
    return _fib(n)[0]


# (Private) Returns the tuple (F(n), F(n+1)).
def _fib(n):
    if n == 0:
        return (0, 1)
    else:
        a, b = _fib(n // 2)
        c = a * (b * 2 - a)
        d = a * a + b * b
        if n % 2 == 0:
            return (c, d)
        else:
            return (d, c + d)
